### Pandas Lab -- Cleaning, Merging, & Grouping

This lab is designed to introduce students to common use cases for Pandas when working with data:

 - Creating new information out of your existing data set
 - Merging, concatenating, and joining different data sources
 - Grouping -- With both time & non-time based data

### Section I: Creating Data Out of Your Existing Columns

Go ahead and create the following columns in your dataset.

**Column 1:**

  - **Column Name:** Profitable
  - **Values:** `True` if `Profit` > 0, `False` if not.

In [25]:
cd DAT07-28-AG/ClassMaterial/Unit2/Data

/Users/imac/DAT07-28-AG/ClassMaterial/Unit2/Data


In [30]:
cd

/Users/imac


In [26]:
import pandas as pd
import numpy as np
# read in the file
df = pd.read_excel("superstore.xls", index_col="Row ID")

In [28]:
df

Order ID Order Date  Ship Date       Ship Mode Customer ID  \
Row ID                                                                     
1       CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
3       CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
4       US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
5       US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
...                ...        ...        ...             ...         ...   
9990    CA-2014-110422 2014-01-21 2014-01-23    Second Class    TB-21400   
9991    CA-2017-121258 2017-02-26 2017-03-03  Standard Class    DB-13060   
9992    CA-2017-121258 2017-02-26 2017-03-03  Standard Class    DB-13060   
9993    CA-2017-121258 2017-02-26 2017-03-03  Standard Class    DB-13060   
9994    CA-2017-119914 2017-05-04 2017-05-09    Second Class    CC-12220   

           Customer Name    Segment        Country             City  \
Row ID                                                                
1            Claire Gute   Consumer  United States        Henderson   
2            Claire Gute   Consumer  United States        Henderson   
3        Darrin Van Huff  Corporate  United States      Los Angeles   
4         Sean O'Donnell   Consumer  United States  Fort Lauderdale   
5         Sean O'Donnell   Consumer  United States  Fort Lauderdale   
...                  ...        ...            ...              ...   
9990    Tom Boeckenhauer   Consumer  United States            Miami   
9991         Dave Brooks   Consumer  United States       Costa Mesa   
9992         Dave Brooks   Consumer  United States       Costa Mesa   
9993         Dave Brooks   Consumer  United States       Costa Mesa   
9994        Chris Cortes   Consumer  United States      Westminster   

             State  ...  Region       Product ID         Category  \
Row ID              ...                                             
1         Kentucky  ...   South  FUR-BO-10001798        Furniture   
2         Kentucky  ...   South  FUR-CH-10000454        Furniture   
3       California  ...    West  OFF-LA-10000240  Office Supplies   
4          Florida  ...   South  FUR-TA-10000577        Furniture   
5          Florida  ...   South  OFF-ST-10000760  Office Supplies   
...            ...  ...     ...              ...              ...   
9990       Florida  ...   South  FUR-FU-10001889        Furniture   
9991    California  ...    West  FUR-FU-10000747        Furniture   
9992    California  ...    West  TEC-PH-10003645       Technology   
9993    California  ...    West  OFF-PA-10004041  Office Supplies   
9994    California  ...    West  OFF-AP-10002684  Office Supplies   

       Sub-Category                                       Product Name  \
Row ID                                                                   
1         Bookcases                  Bush Somerset Collection Bookcase   
2            Chairs  Hon Deluxe Fabric Upholstered Stacking Chairs,...   
3            Labels  Self-Adhesive Address Labels for Typewriters b...   
4            Tables      Bretford CR4500 Series Slim Rectangular Table   
5           Storage                     Eldon Fold 'N Roll Cart System   
...             ...                                                ...   
9990    Furnishings                             Ultra Door Pull Handle   
9991    Furnishings  Tenex B1-RE Series Chair Mats for Low Pile Car...   
9992         Phones                              Aastra 57i VoIP phone   
9993          Paper  It's Hot Message Books with Stickers, 2 3/4" x 5"   
9994     Appliances  Acco 7-Outlet Masterpiece Power Center, Wihtou...   

           Sales  Quantity  Discount    Profit  Profitable  
Row ID                                                      
1       261.9600         2      0.00   41.9136        True  
2       731.9400         3      0.00  219.5820        Tru

In [29]:
df["Profitable"] = np.where(df["Profit"] > 0, True, False)

In [32]:
df

Order ID Order Date  Ship Date       Ship Mode Customer ID  \
Row ID                                                                     
1       CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
3       CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
4       US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
5       US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
...                ...        ...        ...             ...         ...   
9990    CA-2014-110422 2014-01-21 2014-01-23    Second Class    TB-21400   
9991    CA-2017-121258 2017-02-26 2017-03-03  Standard Class    DB-13060   
9992    CA-2017-121258 2017-02-26 2017-03-03  Standard Class    DB-13060   
9993    CA-2017-121258 2017-02-26 2017-03-03  Standard Class    DB-13060   
9994    CA-2017-119914 2017-05-04 2017-05-09    Second Class    CC-12220   

           Customer Name    Segment        Country             City  \
Row ID                                                                
1            Claire Gute   Consumer  United States        Henderson   
2            Claire Gute   Consumer  United States        Henderson   
3        Darrin Van Huff  Corporate  United States      Los Angeles   
4         Sean O'Donnell   Consumer  United States  Fort Lauderdale   
5         Sean O'Donnell   Consumer  United States  Fort Lauderdale   
...                  ...        ...            ...              ...   
9990    Tom Boeckenhauer   Consumer  United States            Miami   
9991         Dave Brooks   Consumer  United States       Costa Mesa   
9992         Dave Brooks   Consumer  United States       Costa Mesa   
9993         Dave Brooks   Consumer  United States       Costa Mesa   
9994        Chris Cortes   Consumer  United States      Westminster   

             State  ...       Product ID         Category Sub-Category  \
Row ID              ...                                                  
1         Kentucky  ...  FUR-BO-10001798        Furniture    Bookcases   
2         Kentucky  ...  FUR-CH-10000454        Furniture       Chairs   
3       California  ...  OFF-LA-10000240  Office Supplies       Labels   
4          Florida  ...  FUR-TA-10000577        Furniture       Tables   
5          Florida  ...  OFF-ST-10000760  Office Supplies      Storage   
...            ...  ...              ...              ...          ...   
9990       Florida  ...  FUR-FU-10001889        Furniture  Furnishings   
9991    California  ...  FUR-FU-10000747        Furniture  Furnishings   
9992    California  ...  TEC-PH-10003645       Technology       Phones   
9993    California  ...  OFF-PA-10004041  Office Supplies        Paper   
9994    California  ...  OFF-AP-10002684  Office Supplies   Appliances   

                                             Product Name     Sales Quantity  \
Row ID                                                                         
1                       Bush Somerset Collection Bookcase  261.9600        2   
2       Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
3       Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
4           Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
5                          Eldon Fold 'N Roll Cart System   22.3680        2   
...                                                   ...       ...      ...   
9990                               Ultra Door Pull Handle   25.2480        3   
9991    Tenex B1-RE Series Chair Mats for Low Pile Car...   91.9600        2   
9992                                Aastra 57i VoIP phone  258.5760        2   
9993    It's Hot Message Books with Stickers, 2 3/4" x 5"   29.6000        4   
9994    Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.1600        2   

        Discount    Profit  Profitable  Actual Ship Time  
Row ID                                

**Column 2:**

 - **Column Name:** Expected Ship Time
 - **Values:**
   - `0` if `Ship Mode` == `Same Day`
   - `2` if `Ship Mode` == `First Class`
   - `3` if `Ship Mode` == `Second Class`
   - `6` if `Ship Mode` == `Standard Class`
   - `-1` if none of the above.

In [35]:
conditions = [
    df["Ship Mode"] == "Same Day",
    df["Ship Mode"] == "First Class",
    df["Ship Mode"] == "Second Class",
    df["Ship Mode"] == "Standard Class"
]

results = [0, 2, 3, 6]

df["Expected Ship Time"] = np.select(conditions, results, -1)

**Column 3:**

 - **Column Name:** Actual Ship Time
 - **Values:**
   - `Ship Date` - `Order Date`
 - **Note:** When you subtract these columns, your column will be a **time delta**.  See if you can use the `dt` attribute to convert these values into an integer.  Ie, if your value reads `3 days`, you want that to be 3 instead.  You can read more about different time periods in pandas here:  https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#time-date-components

In [36]:
df["Actual Ship Time"] = (df["Ship Date"] - df["Order Date"]).dt.days

**Column 4:**

 - **Column Name:** Late
 - **Values:** `True` if `Actual Ship Time` > `Expected Ship Time`, `False` otherwise

In [37]:
df["Late"] = (df["Actual Ship Time"] > df["Expected Ship Time"])

In [38]:
df

Order ID Order Date  Ship Date       Ship Mode Customer ID  \
Row ID                                                                     
1       CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
3       CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
4       US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
5       US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
...                ...        ...        ...             ...         ...   
9990    CA-2014-110422 2014-01-21 2014-01-23    Second Class    TB-21400   
9991    CA-2017-121258 2017-02-26 2017-03-03  Standard Class    DB-13060   
9992    CA-2017-121258 2017-02-26 2017-03-03  Standard Class    DB-13060   
9993    CA-2017-121258 2017-02-26 2017-03-03  Standard Class    DB-13060   
9994    CA-2017-119914 2017-05-04 2017-05-09    Second Class    CC-12220   

           Customer Name    Segment        Country             City  \
Row ID                                                                
1            Claire Gute   Consumer  United States        Henderson   
2            Claire Gute   Consumer  United States        Henderson   
3        Darrin Van Huff  Corporate  United States      Los Angeles   
4         Sean O'Donnell   Consumer  United States  Fort Lauderdale   
5         Sean O'Donnell   Consumer  United States  Fort Lauderdale   
...                  ...        ...            ...              ...   
9990    Tom Boeckenhauer   Consumer  United States            Miami   
9991         Dave Brooks   Consumer  United States       Costa Mesa   
9992         Dave Brooks   Consumer  United States       Costa Mesa   
9993         Dave Brooks   Consumer  United States       Costa Mesa   
9994        Chris Cortes   Consumer  United States      Westminster   

             State  ...  Sub-Category  \
Row ID              ...                 
1         Kentucky  ...     Bookcases   
2         Kentucky  ...        Chairs   
3       California  ...        Labels   
4          Florida  ...        Tables   
5          Florida  ...       Storage   
...            ...  ...           ...   
9990       Florida  ...   Furnishings   
9991    California  ...   Furnishings   
9992    California  ...        Phones   
9993    California  ...         Paper   
9994    California  ...    Appliances   

                                             Product Name     Sales Quantity  \
Row ID                                                                         
1                       Bush Somerset Collection Bookcase  261.9600        2   
2       Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
3       Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
4           Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
5                          Eldon Fold 'N Roll Cart System   22.3680        2   
...                                                   ...       ...      ...   
9990                               Ultra Door Pull Handle   25.2480        3   
9991    Tenex B1-RE Series Chair Mats for Low Pile Car...   91.9600        2   
9992                                Aastra 57i VoIP phone  258.5760        2   
9993    It's Hot Message Books with Stickers, 2 3/4" x 5"   29.6000        4   
9994    Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.1600        2   

       Discount    Profit  Profitable  Actual Ship Time  Expected Ship Time  \
Row ID                                                                        
1          0.00   41.9136        True                 3                   3   
2          0.00  219.5820        True                 3                   3   
3          0.00    6.8714        True                 4                   3   
4          0.45 -383.0310       False                 7                   6   
5          0.20    2.5164        True               

### Section II: Merging Dataframes

This excel spreadsheet has 3 separate sheets.  Look up the documentation on the `pd.read_excel` method on how to load in the other two.  

After that, merge the other two dataframes into your original one, and make sure your original dataset now has the following columns:

 - **Salesperson:** This is the Salesperson in charge of each region.
 - **Returned:** This details whether or not the order was returned.  Fill in null values with the value `no`.
 
Use the `drop()` method if you need to get rid of redundant columns.

**Important:** We want to keep all of the rows in the dataset we first loaded in.  After each merge, it's a good idea to make sure your dataset hasn't shrunk, which will happen if you don't choose the right merge type.  Make sure you have 9,994 rows when you're finished!

In [42]:
df2 = pd.read_excel("/Users/imac/DAT07-28-AG/ClassMaterial/Unit2/Data/superstore.xls", sheet_name=1)

In [41]:
df3 = pd.read_excel("/Users/imac/DAT07-28-AG/ClassMaterial/Unit2/Data/superstore.xls", sheet_name=2)

In [43]:
df = df.merge(df2, on="Order ID", how="left")

In [44]:
df["Returned"].value_counts(dropna=False)

NaN    9194
Yes     800
Name: Returned, dtype: int64

In [45]:
df["Returned"].fillna("No", inplace=True)

In [46]:
df = df.merge(df3, on="Region", how="inner")

In [48]:
df.head()

Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
3  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4  CA-2017-114412 2017-04-15 2017-04-20  Standard Class    AA-10480   

    Customer Name   Segment        Country             City           State  \
0     Claire Gute  Consumer  United States        Henderson        Kentucky   
1     Claire Gute  Consumer  United States        Henderson        Kentucky   
2  Sean O'Donnell  Consumer  United States  Fort Lauderdale         Florida   
3  Sean O'Donnell  Consumer  United States  Fort Lauderdale         Florida   
4    Andrew Allen  Consumer  United States          Concord  North Carolina   

   ...     Sales Quantity Discount    Profit Profitable Actual Ship Time  \
0  ...  261.9600        2     0.00   41.9136       True                3   
1  ...  731.9400        3     0.00  219.5820       True                3   
2  ...  957.5775        5     0.45 -383.0310      False                7   
3  ...   22.3680        2     0.20    2.5164       True                7   
4  ...   15.5520        3     0.20    5.4432       True                5   

   Expected Ship Time   Late  Returned             Person  
0                   3  False        No  Cassandra Brandow  
1                   3  False        No  Cassandra Brandow  
2                   6   True        No  Cassandra Brandow  
3                   6   True        No  Cassandra Brandow  
4                   6  False        No  Cassandra Brandow  

[5 rows x 26 columns]

In [49]:
df.rename({"Person": "Salesperson"}, axis=1, inplace=True)

In [50]:
df.head()

Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
3  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4  CA-2017-114412 2017-04-15 2017-04-20  Standard Class    AA-10480   

    Customer Name   Segment        Country             City           State  \
0     Claire Gute  Consumer  United States        Henderson        Kentucky   
1     Claire Gute  Consumer  United States        Henderson        Kentucky   
2  Sean O'Donnell  Consumer  United States  Fort Lauderdale         Florida   
3  Sean O'Donnell  Consumer  United States  Fort Lauderdale         Florida   
4    Andrew Allen  Consumer  United States          Concord  North Carolina   

   ...     Sales Quantity Discount    Profit Profitable Actual Ship Time  \
0  ...  261.9600        2     0.00   41.9136       True                3   
1  ...  731.9400        3     0.00  219.5820       True                3   
2  ...  957.5775        5     0.45 -383.0310      False                7   
3  ...   22.3680        2     0.20    2.5164       True                7   
4  ...   15.5520        3     0.20    5.4432       True                5   

   Expected Ship Time   Late  Returned        Salesperson  
0                   3  False        No  Cassandra Brandow  
1                   3  False        No  Cassandra Brandow  
2                   6   True        No  Cassandra Brandow  
3                   6   True        No  Cassandra Brandow  
4                   6  False        No  Cassandra Brandow  

[5 rows x 26 columns]

### Section III: Grouping

Use the `groupby` or `resample` method to answer the following questions.

**Question 1:** What salesperson had the highest average sales amount? 

In [51]:
df.groupby("Salesperson")["Sales"].mean()

Salesperson
Anna Andreadi        226.493233
Cassandra Brandow    241.803645
Chuck Magee          238.336110
Kelly Williams       215.772661
Name: Sales, dtype: float64

**Question 2:** Within each ship mode, compare how likely late orders were to be profitable or not

In [53]:
df.groupby(["Ship Mode", "Late"])["Profitable"].mean()

Ship Mode       Late 
First Class     False    0.795405
                True     0.814103
Same Day        False    0.813102
                True     0.916667
Second Class    False    0.845415
                True     0.827500
Standard Class  False    0.794651
                True     0.805153
Name: Profitable, dtype: float64

**Question 3:** What ship mode had the most consistently on time orders?

In [55]:
df.groupby("Ship Mode")["Late"].mean()

Ship Mode
First Class       0.405722
Same Day          0.044199
Second Class      0.411311
Standard Class    0.104055
Name: Late, dtype: float64

**Question 4:** For each sales person, get their average, median, max, and count of their sales.

In [56]:
df.groupby("Salesperson")["Sales"].agg(["mean", "median", "max", "count"])

mean  median        max  count
Salesperson                                            
Anna Andreadi      226.493233  60.840  13999.960   3203
Cassandra Brandow  241.803645  54.594  22638.480   1620
Chuck Magee        238.336110  54.900  11199.968   2848
Kelly Williams     215.772661  45.980  17499.950   2323

**Question 5:** Group your dataset according to  `Region`, and `Category`, and then call the `describe()` method to get the summary statistics for each subgroup.

In [57]:
df.groupby(["Region", "Category"]).describe()

Postal Code                                       \
                              count          mean           std      min   
Region  Category                                                           
Central Furniture             481.0  65486.453222  11200.602622  46060.0   
        Office Supplies      1422.0  65275.734880  11481.099454  46060.0   
        Technology            420.0  65393.026190  11530.132828  46060.0   
East    Furniture             601.0  17317.247920  12542.460289   1040.0   
        Office Supplies      1712.0  17401.548481  12657.699221   1453.0   
        Technology            535.0  18640.338318  13202.282930   1841.0   
South   Furniture             332.0  34552.117470  10839.698410  22153.0   
        Office Supplies       995.0  34243.389950  10595.433133  22153.0   
        Technology            293.0  35261.351536  12440.300306  22153.0   
West    Furniture             707.0  91680.442716   5024.265567  59601.0   
        Office Supplies      1897.0  91769.716921   5180.811997  59102.0   
        Technology            599.0  91543.449082   5473.761207  59405.0   

                                                              Sales  \
                             25%      50%      75%      max   count   
Region  Category                                                      
Central Furniture        56301.0  62521.0  77041.0  79907.0   481.0   
        Office Supplies  55122.0  65807.0  77041.0  79907.0  1422.0   
        Technology       55083.5  65807.0  77041.0  79907.0   420.0   
East    Furniture        10011.0  11561.0  19143.0  45503.0   601.0   
        Office Supplies  10011.0  11535.0  19143.0  45503.0  1712.0   
        Technology       10011.0  13501.0  19711.0  45503.0   535.0   
South   Furniture        28314.0  33023.5  37918.0  72701.0   332.0   
        Office Supplies  28403.0  32712.0  37211.0  72762.0   995.0   
        Technology       28403.0  32216.0  37211.0  72701.0   293.0   
West    Furniture        90032.0  92105.0  94521.0  99301.0   707.0   
        Office Supplies  90032.0  92374.0  94513.0  99301.0  1897.0   
        Technology       90008.0  92037.0  94513.0  99207.0   599.0   

                                     ... Actual Ship Time       \
                               mean  ...              75%  max   
Region  Category                     ...                         
Central Furniture        340.534644  ...              5.0  7.0   
        Office Supplies  117.458801  ...              5.0  7.0   
        Technology       405.753124  ...              5.0  7.0   
East    Furniture        346.574383  ...              5.0  7.0   
        Office Supplies  120.044425  ...              5.0  7.0   
        Technology       495.278469  ...              5.0  7.0   
South   Furniture        353.309289  ...              5.0  7.0   
        Office Supplies  126.282727  ...              5.0  7.0   
        Technology       507.753952  ...              5.0  7.0   
West    Furniture        357.302325  ...              5.0  7.0   
        Office Supplies  116.422377  ...              5.0  7.0   
        Technology       420.687533  ...              5.0  7.0   

                        Expected Ship Time                                     \
                                     count      mean       std  min  25%  50%   
Region  Category                                                                
Central Furniture                    481.0  4.532225  1.958821  0.0  3.0  6.0   
        Office Supplies             1422.0  4.626582  1.913822  0.0  3.0  6.0   
        Technology                   420.0  4.447619  1.899534  0.0  3.0  6.0   
East    Furniture                    601.0  4.377704  1.982112  0.0  3.0  6.0   
        Office Supplies             1712.0  4.438668  1.972867  0.0  3.0  6.0   
        Technology                   535.0  4.444860  1.994432  0.0  3.0  6.0   
South   Furniture                    332.0  4.533133  1.895005  0.0  3.0  6.0   
        Office S

**Question 6:** Use the `Resample()` method to get the sum of sales for each quarter.

In [60]:
df.resample("Q", on="Order Date")["Sales"].sum()

Order Date
2014-03-31     74447.7960
2014-06-30     86538.7596
2014-09-30    143633.2123
2014-12-31    179627.7302
2015-03-31     68851.7386
2015-06-30     89124.1870
2015-09-30    130259.5752
2015-12-31    182297.0082
2016-03-31     93237.1810
2016-06-30    136082.3010
2016-09-30    143787.3622
2016-12-31    236098.7538
2017-03-31    123144.8602
2017-06-30    133764.3720
2017-09-30    196251.9560
2017-12-31    280054.0670
Freq: Q-DEC, Name: Sales, dtype: float64

**Question 7:** What quarter had the highest total sales amount?

In [62]:
df.resample("Q", on="Order Date")["Sales"].sum().max()

280054.067

In [70]:
df.resample("Q", on="Order Date")["Sales"].sum().idxmax()

Timestamp('2017-12-31 00:00:00', freq='Q-DEC')

**Question 8:** See if you can use the `groupby` method to get a list of yearly sales for each region inside the dataset.

**Hint:** Try using the `dt` attribute of the `Order Date` column.

In [71]:
df.groupby(["Region", df["Order Date"].dt.year])["Sales"].sum()

Region   Order Date
Central  2014          103838.1646
         2015          102874.2220
         2016          147429.3760
         2017          147098.1282
East     2014          128680.4570
         2015          156332.0570
         2016          180685.8220
         2017          213082.9040
South    2014          103845.8435
         2015           71359.9805
         2016           93610.2235
         2017          122905.8575
West     2014          147883.0330
         2015          139966.2495
         2016          187480.1765
         2017          250128.3655
Name: Sales, dtype: float64

**Bonus:** Creating summary statistics with a `groupby` statement.  

Lots of times it's very useful to be able to create a summary statistic for a particular category to compare with individual samples.

For example, if you were doing fraud detection, and someone were making a purchase at a 7-11, a transaction amount of $175 would be unusually large for someone making a purchase at a convenience store, essentially setting off a red flag that the transaction might be suspicious.

Creating such comparisons is easily done using the `groupby` method and then merging it back into the original dataframe.  

For example, if you wanted to compare every single purchase amount with the average amount for that category, you could do it in the following way:

In [2]:
import pandas as pd
df = pd.read_excel('../data/superstore.xls')
# create the grouping
cat_grouping = df.groupby('Category')[['Sales']].mean()
# this step is mostly just to make the merged dataframe more tidy
cat_grouping.rename({'Sales': 'Cat_Average'}, axis=1, inplace=True)

In [3]:
# join them
df = df.merge(cat_grouping, left_on='Category', right_index=True)

In [4]:
# and now we can see each purchase amount compared to the average amt
# for that category
df.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0        1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1        2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
3        4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
5        6  CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
10      11  CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   

      Customer Name   Segment        Country             City     ...       \
0       Claire Gute  Consumer  United States        Henderson     ...        
1       Claire Gute  Consumer  United States        Henderson     ...        
3    Sean O'Donnell  Consumer  United States  Fort Lauderdale     ...        
5   Brosina Hoffman  Consumer  United States      Los Angeles     ...        
10  Brosina Hoffman  Consumer  United States      Los Angeles     ...        

   Region       Product ID   Category Sub-Category  \
0   South  FUR-BO-10001798  Furniture    Bookcases   
1   South  FUR-CH-10000454  Furniture       Chairs   
3   South  FUR-TA-10000577  Furniture       Tables   
5    West  FUR-FU-10001487  Furniture  Furnishings   
10   West  FUR-TA-10001539  Furniture       Tables   

                                         Product Name      Sales Quantity  \
0                   Bush Somerset Collection Bookcase   261.9600        2   
1   Hon Deluxe Fabric Upholstered Stacking Chairs,...   731.9400        3   
3       Bretford CR4500 Series Slim Rectangular Table   957.5775        5   
5   Eldon Expressions Wood and Plastic Desk Access...    48.8600        7   
10           Chromcraft Rectangular Conference Tables  1706.1840        9   

    Discount    Profit  Cat_Average  
0       0.00   41.9136   349.834887  
1       0.00  219.5820   349.834887  
3       0.45 -383.0310   349.834887  
5       0.00   14.1694   349.834887  
10      0.20   85.3092   349.834887  

[5 rows x 22 columns]

So for instance, if wanted to ask ourselves, "which customers consistently punch above their weight when it comes to the actual items that they buy?"

We could easily do the following:

In [5]:
# turn the difference between the two columns into a percent
df['Cat Difference'] = ((df['Sales'] / df['Cat_Average']) - 1) * 100

In [6]:
# now group and sort the values
df.groupby('Customer Name')['Cat Difference'].mean().sort_values(ascending=False)

Customer Name
Mitch Willingham        845.569285
Christopher Martinez    636.313301
Andy Reiter             449.654450
Adrian Barton           418.548520
Sanjit Chand            386.906876
Amy Cox                 363.031889
Yoseph Carroll          344.794116
Yana Sorensen           317.572571
Sean Miller             308.018602
Tamara Chand            304.928500
Alex Avila              296.561353
Greg Maxwell            292.917810
Grant Thornton          291.980410
Jane Waco               272.432920
Tom Ashbrook            270.172385
Paul Knutson            244.824972
Robert Dilbeck          239.904239
Gary Hwang              231.081468
Ken Lonsdale            216.273514
Dennis Pardue           213.919098
Justin Hirsh            205.806859
Stefanie Holloman       199.625920
Cathy Prescott          194.089483
Bill Shonely            194.068945
Erica Smith             193.504802
Neil Ducich             174.357173
Adam Bellavance         174.177216
Hunter Lopez            170.364451
Karen 

**Your Turn:** Using a similar methodology as above, figure out the 10 customers who are the most profitable on average, when compared to the subcategory that they purchased from.

If you wanted, you could also limit this to customers who only made a minimum number of purchases as well.

In [ ]:
# your answer here